src/classes/Collision/CapsuleConvexHull.js

```diff
@@ -0,0 +1,240 @@
+        if ( Math.abs( dotBetweenInnerSegmentDirectionAndContactNormal ) > epsilon ) {
+            Goblin.Collision.capsuleConvexHull._fillContactWithDetails( contact, closestPointOnInnerSegment, closestPointOnHull );
+            return [ contact ];
+        }
+
+        invertedRotationQuaternion.invertQuaternion( convexHull.rotation );
+        invertedRotationQuaternion.transformVector3Into( contactNormal, referencePlaneNormal );
+        referencePlaneNormal.scale( -1 );
+        var referenceFace = null;
+        for ( var i = 0; i < convexHull.shape.faces.length; i++ ) {
+            var face = convexHull.shape.faces[ i ];
+            // Let's take a moment and pray for convex hull building algorithm to successfully merge all faces with the same normal.
+            if ( face.normal.equals( referencePlaneNormal, epsilon ) ) {
+                referenceFace = face;
+                break;
+            }
+        }
+
+        if ( referenceFace === null ) {
+            Goblin.Collision.capsuleConvexHull._fillContactWithDetails( contact, closestPointOnInnerSegment, closestPointOnHull );
+            return [ contact ];
+        }
+
+        var clippingWasSuccessful = Goblin.Collision.capsuleConvexHull._clipLineSegmentWithFace(
+            innerSegmentStart,
+            innerSegmentEnd,
+            referenceFace,
+            clippedInnerSegmentStart,
+            clippedInnerSegmentEnd
+        );
+        if ( !clippingWasSuccessful ) {
+            Goblin.Collision.capsuleConvexHull._fillContactWithDetails( contact, closestPointOnInnerSegment, closestPointOnHull );
+            return [ contact ];
+        }
+
+        var contactAtTheStart = contact;
+        var contactAtTheEnd = Goblin.ObjectPool.getObject( 'ContactDetails' );
+        contactAtTheEnd.object_a = capsule;
+        contactAtTheEnd.object_b = convexHull;
+        contactAtTheEnd.contact_normal.copy( contactAtTheStart.contact_normal );
+        contactAtTheEnd.penetration_depth = contactAtTheStart.penetration_depth;
+
+        contactAtTheStart.contact_point_in_b.scaleVector( contactNormal, capsule.shape.radius - penetrationDepth );
+        contactAtTheStart.contact_point_in_b.add( clippedInnerSegmentStart );
+        contactAtTheEnd.contact_point_in_b.scaleVector( contactNormal, capsule.shape.radius - penetrationDepth );
+        contactAtTheEnd.contact_point_in_b.add( clippedInnerSegmentEnd );
+
+        Goblin.Collision.capsuleConvexHull._fillContactWithDetails( contactAtTheStart, clippedInnerSegmentStart, contactAtTheStart.contact_point_in_b );
+        Goblin.Collision.capsuleConvexHull._fillContactWithDetails( contactAtTheEnd, clippedInnerSegmentEnd, contactAtTheEnd.contact_point_in_b );
+
+        return [ contactAtTheStart, contactAtTheEnd ];
+    };
+} )();
+
+Goblin.Collision.capsuleConvexHull._deepCapsuleConvexHull = ( function() {
+    return function( capsule, convexHull, doLightweightCollision ) {
+        /**
```

Comment:
Are you absolutely sure we need these comments? I mean it's kinda obvious which object is returned, and, in worst case, we could use a normal // comment instead to note take 3 lines for a type hint.

Reply:
I just like a little bit of the strong typings / type hints. However, I 100% agree that such comments take up a lot of space. Will remove